### PR TITLE
Pr voxtlm

### PR DESCRIPTION
**Pull request by soumimaiti**
## What?

Adding new task template for Language Model.  

## Why?

Allow training for language models and adding recipe.

## See also

We are using asr1 template and modifying to keep language model as a separate task. Additionally we can also train both speech and text joint language model using voxtlm (https://arxiv.org/pdf/2309.07937.pdf) .
We will use this template to create new recipe for training voxtlm.

### CODE REVIEW

File: egs2/TEMPLATE/lm1/lm.sh
Comment (sw005320) [inline egs2/TEMPLATE/lm1/lm.sh:1]: Please complete this.

**Review by sw005320 — COMMENTED** [review]
Please check the other PR (e.g., asr2 PR https://github.com/espnet/espnet/pull/5181) and where we should add.
The template part should be simpler as it only performs LM training, but the recipe part would be a bit complicated due to the mix of speech and text corpora.

**Comment by sw005320** [comment]
@soumimaiti, please finish this PR as soon as possible.

@pyf98, @simpleoier, @wyh2000, let’s help @soumimaiti to finish this PR quickly.

**Comment by sw005320** [comment]
Can you start to add a recipe as well?

**Comment by sw005320** [comment]
Thx @soumimaiti for the update.
There are some CI issues https://github.com/espnet/espnet/actions/runs/6672890458/job/18137601837?pr=5472#step:8:311
Please fix them.

@pyf98, @wyh2000, @simpleoier, please start the reviews.

File: egs2/TEMPLATE/asr1/db.sh [hunk 1]
```diff
@@ -53,8 +53,9 @@ HUI_ACG=downloads
 HUB4_SPANISH=
 LABOROTV=
 TEDXJP=
-LIBRISPEECH=downloads
+LIBRISPEECH="/swl/projects/corpora/LibriSpeech"
```
Comment (pyf98) [inline egs2/TEMPLATE/asr1/db.sh:56]: Please restore the value to default

File: egs2/TEMPLATE/asr1/db.sh [hunk 2]
```diff
@@ -69,7 +70,7 @@ MISP2021=
 MLSUPERB=
 L3DAS22=
 LIBRIMIX=downloads
-LIBRITTS=downloads
+LIBRITTS=/swl/projects/corpora/LibriTTS
```
Comment (pyf98) [inline egs2/TEMPLATE/asr1/db.sh:73]: Same here

File: egs2/TEMPLATE/lm1/lm.sh
Comment (sw005320) [inline egs2/TEMPLATE/lm1/lm.sh:1]: Can you add the scoring parts?
I think this part is not easy, but it makes the recipe more attractive.

File: egs2/voxtlm/lm1/local/prepare_bpe_text.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import argparse
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-i", "--inp", type=str, help="Input file with all training data"
```
Comment (sw005320) [inline egs2/voxtlm/lm1/local/prepare_bpe_text.py:6]:
```suggestion
        "-i", "--input", type=str, help="Input file with all training data"
```
We usually use more complete words or phrases for `--` based options

File: egs2/voxtlm/lm1/local/prepare_bpe_text.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    )
+    parser.add_argument("-o", "--out", type=str, help="Output file for bpe training")
```
Comment (sw005320) [inline egs2/voxtlm/lm1/local/prepare_bpe_text.py:8]:
```suggestion
    parser.add_argument("-o", "--output", type=str, help="Output file for bpe training")
```
ditto
We usually use more complete words or phrases for `--` based options

File: egs2/voxtlm/lm1/local/prepare_bpe_text.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    parser.add_argument(
+        "-n",
+        "--num",
```
Comment (sw005320) [inline egs2/voxtlm/lm1/local/prepare_bpe_text.py:11]: ditto
```suggestion
        "--num-utterances",
```

File: egs2/voxtlm/lm1/local/prepare_lm_data.py
Comment (sw005320) [inline egs2/voxtlm/lm1/local/prepare_lm_data.py:1]: Can you make the special symbols as a variable?
You can make an option to specify it (but make the current ones as default).

File: egs2/voxtlm/lm1/local/prepare_lm_test.py
Comment (sw005320) [inline egs2/voxtlm/lm1/local/prepare_lm_test.py:1]: ditto
Can you make the special symbols as a variable?
You can make an option to specify it (but make the current ones as default).

File: egs2/voxtlm/lm1/local/text_normalize.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    exit 1
+fi
+
+# remove punctuations from text
+if [ "$#" -ne 1 ]; then
+    echo "Illegal number of parameters"
+fi
+
+text_dir=$1
+
```
Comment (sw005320) [inline egs2/voxtlm/lm1/local/text_normalize.sh:19]: I recommend you show an example before/after this normalization here.

File: espnet2/lm/espnet_model.py
Comment (sw005320) [inline espnet2/lm/espnet_model.py:1]: - Can you remove the comment out lines?
- I'm not sure this change does not affect the other parts. Please add a test to verify it.

**Review by sw005320 — COMMENTED** [review]
Please also start to add the unit test and integration test.
@ftshijt, can you help @soumimaiti for this?

**Comment by sw005320** [comment]
> I just have a general suggestion: Let's try to keep the original LM definition unchanged as much as possible. Otherwise it might interfere with our standard ASR task, which has not been thoroughly investigated under this PR. If we need any new features that are specific to VoxtLM, we can create new LM model classes by inheritance. This is also easier to extend in the future.

I agree.
Considering various extension of VoxtLM, making a new model class in this stage is a good idea.

**Comment by sw005320** [comment]
Please also fix the CI error.

File: espnet2/lm/multitask_espnet_model.py
Comment (sw005320) [inline espnet2/lm/multitask_espnet_model.py:1]: Can you rename it `espnet_model_multitask.py`?
The other examples use `espnet_model` as a prefix (https://github.com/espnet/espnet/tree/master/espnet2/enh)

File: egs2/TEMPLATE/asr1/db.sh (hunk 1 above)
Comment (soumimaiti) [reply egs2/TEMPLATE/asr1/db.sh]: Restored.

File: egs2/voxtlm/lm1/local/prepare_lm_data.py
Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/prepare_lm_data.py:1]: Added as variable.

File: egs2/voxtlm/lm1/local/prepare_lm_test.py
Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/prepare_lm_test.py:1]: Added as variable.

File: egs2/voxtlm/lm1/local/text_normalize.sh (hunk 6 above)
Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/text_normalize.sh:21]: Added example.

File: espnet2/lm/espnet_model.py
Comment (soumimaiti) [reply espnet2/lm/espnet_model.py:1]: Now added a new model class for VoxtLM. Will add a test.

File: egs2/voxtlm/lm1/local/prepare_lm_data.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    uttid2text = {}
+    with text.open("r") as fp:
+        for line in fp.readlines():
+            line = line.strip().split()
+            uttid2text[line[0]] = " ".join(line[1:])
+    return uttid2text
+
+
+# Convert discrete unit to CJK characters
+# CJK Unified Ideographs (20,977 chars): \u4e00 - \u9fff
```
Comment (ftshijt) [inline egs2/voxtlm/lm1/local/prepare_lm_data.py:18]: Quick question here on whether the usage of CJK token would be good: my understanding for CJK is its easiness of compute BPE, however we have following constraints/issues
- cluster size (limits to a boundary)
- worse readability (as the CJK mapping is not quite obvious)
- meaningless for multi-stream modeling of speech: as for multi-stream (e.g., neural codec etc) we cannot achieve bpe on each stream, thus cannot gain from the benefits of achieving BPE easily.

Please let me know if you have other thoughts~ I guess the decision here would be very fundamental as it will also affects the modeling in many other tasks:
- in s2st and svs2, we directly use text_int (sequence of integer)
- in asr2 and st2, we use cjk to reduce sequence
- we are still in discussion of which format to use in tts2

Comment (ftshijt) [reply egs2/voxtlm/lm1/local/prepare_lm_data.py:18]: For another comment on the design: whether we would want the cjk conversion in local or within a certain stage?

File: egs2/voxtlm/lm1/local/librispeech/download_and_untar.sh [hunk 8]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/librispeech/download_and_untar.sh:1]: For Librilight we don't have a recipe yet. We can use symlink for datasets, main reason for using separate script is to not have the dependency on other recipes if something changes.

File: egs2/voxtlm/lm1/local/prepare_lm_data.py (hunk 7 above)
Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/prepare_lm_data.py:18]: Since we do extraction of data on stage 2 and stage 3, we have to convert cjk after or within stage 3. But I agree we can do that outside local as well.

Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/prepare_lm_data.py:18]: Maybe we can add the CJK token mapping with BPE, as it is needed for BPE.

Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/prepare_lm_data.py:18]: Added CJK token as a parameter, can be turned off depending on recipe.

File: espnet2/lm/espnet_model_multitask.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        # nll: (BxL,)
+        nll = F.cross_entropy(y.view(-1, y.shape[-1]), t.view(-1), reduction="none")
+        # nll: (BxL,) -> (BxL,)
+        if max_length is None:
+            nll.masked_fill_(make_pad_mask(x_lengths).to(nll.device).view(-1), 0.0)
+        else:
+            # nll.masked_fill_(
+            # make_pad_mask(x_lengths, maxlen=max_length + 1).to(nll.device).view(-1),
+            #     0.0,
+            # )
```
Comment (sw005320) [inline espnet2/lm/espnet_model_multitask.py:97]: Do you want to keep them?
If so, please leave some comments of why do you want to keep them.

**Comment by sw005320** [comment]
I think it's getting a better shape.
We need more tests, integration tests, and results in the recipe (README.md).
We may also be careful about the name of recipes (voxtlm is the name of the method. Not a recipe name)

@ftshijt, @pyf98, @simpleoier, can you do the second round of reviews?

File: espnet2/lm/espnet_model_multitask.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor], torch.Tensor]:
+        batch_size = text.shape[0]
+        loss, acc = self._calc_att_loss(text, text_lengths)
+        stats = dict(
+            loss=loss.detach(),
+            acc=acc,
+        )
+
+        # force_gatherable: to-device and to-tensor if scalar for DataParallel
+        # loss, stats, weight = force_gatherable((loss, stats, ntokens), loss.device)
```
Comment (pyf98) [inline espnet2/lm/espnet_model_multitask.py:198]: Do we need this commented line?

File: egs2/TEMPLATE/asr1/db.sh [hunk 11]
```diff
@@ -59,6 +59,7 @@ LABOROTV=
 TEDXJP=
 LIBRISPEECH=downloads
 LIBRILIGHT_LIMITED=
+LIBRILIGHT="downloads"
```
Comment (pyf98) [inline egs2/TEMPLATE/asr1/db.sh:62]: "downloads" means the dataset can be automatically downloaded by the data.sh script. Do we support this auto-downloading for librilight?

File: egs2/voxtlm/lm1/local/data_librilight.sh
Comment (pyf98) [inline egs2/voxtlm/lm1/local/data_librilight.sh:1]: This script has a lot of comments. Can we clean it?

**Comment by sw005320** [comment]
@soumimaiti, please continue working on it. It's getting better and better. Also, I have several requests for voxtlm.

**Comment by soumimaiti** [comment]
> I think it's getting a better shape. We need more tests, integration tests, and results in the recipe (README.md). We may also be careful about the name of recipes (voxtlm is the name of the method. Not a recipe name)
> 
> @ftshijt, @pyf98, @simpleoier, can you do the second round of reviews?

I agree about the name.  Do we follow any strategy for recipe with multiple datasets ? Any suggestion would be useful.

**Comment by sw005320** [comment]
> > I think it's getting a better shape. We need more tests, integration tests, and results in the recipe (README.md). We may also be careful about the name of recipes (voxtlm is the name of the method. Not a recipe name)
> > @ftshijt, @pyf98, @simpleoier, can you do the second round of reviews?
> 
> I agree about the name. Do we follow any strategy for recipe with multiple datasets ? Any suggestion would be useful.

Like owsm, how about adding the version, e.g., `voxtlm_v1`?
Also, if you extend it to multilingual, it would be a good idea to add the number of languages (e.g., `opne_li52`)

File: espnet2/lm/espnet_model_multitask.py (hunk 9 above)
Comment (soumimaiti) [reply espnet2/lm/espnet_model_multitask.py]: Removed

File: espnet2/lm/espnet_model_multitask.py (hunk 10 above)
Comment (soumimaiti) [reply espnet2/lm/espnet_model_multitask.py]: Removed

File: egs2/voxtlm/lm1/local/data_librilight.sh
Comment (soumimaiti) [reply egs2/voxtlm/lm1/local/data_librilight.sh:1]: Cleaned.

**Comment by sw005320** [comment]
I'm not sure `voxtlm_en` is a reasonable name.
There are a lot of English speech corpora, and this information is ambiguous. Also, if we change the data (e.g., more English data), how to deal with it?
Can you reconsider the name?

**Comment by sw005320** [comment]
Can you add a test for the training part?
You may refer to https://github.com/espnet/espnet/blob/master/test/espnet2/bin/test_lm_train.py

**Review by sw005320 — COMMENTED** [review]
- please fix the CI errors: https://github.com/espnet/espnet/actions/runs/7614746170/job/20737731263?pr=5472#step:8:1049
- please work on the scoring part in later PRs.

File: egs2/mini_an4/lm1/local/data_bk.sh
[the remark this answers is not in the document]
Comment (soumimaiti) [reply egs2/mini_an4/lm1/local/data_bk.sh:1]: Cleaned

File: egs2/mini_an4/lm1/local/data_prep_bk
[the remark this answers is not in the document]
Comment (soumimaiti) [reply egs2/mini_an4/lm1/local/data_prep_bk:1]: Cleaned

File: egs2/mini_an4/lm1/lm2.sh
[the remark this answers is not in the document]
Comment (soumimaiti) [reply egs2/mini_an4/lm1/lm2.sh:1]: cleaned

**Comment by sw005320** [comment]
I think there is an issue in the integration test, but I could not identify it.
https://github.com/espnet/espnet/actions/runs/7707664847/job/21283480881?pr=5472
Can you check it?

**Comment by wyh2000** [comment]
> I think there is an issue in the integration test, but I could not identify it. https://github.com/espnet/espnet/actions/runs/7707664847/job/21283480881?pr=5472 Can you check it?

This issue was fixed

**Comment by sw005320** [comment]
Thanks, @soumimaiti and @wyh2000!
Now I merge this PR!
